### PR TITLE
Add [Shake to summarize] FXIOS-13145 title animation

### PR DESCRIPTION
**Pull request by FilippoZazzeroni**
## :scroll: Tickets
[Jira ticket](https://mozilla-hub.atlassian.net/browse/FXIOS-13145)
[Github issue](https://github.com/mozilla-mobile/firefox-ios/issues/28599)

## :bulb: Description
Add title animation and refactor summary view in its own component.

## Glass 
https://github.com/user-attachments/assets/766b5808-80c2-44a1-9ff7-0feedd4cbbd3

## Fallback
https://github.com/user-attachments/assets/d4199ee9-81a8-48d6-9208-3ecdf6606cc6

## :pencil: Checklist
- [x] I filled in the ticket numbers and a description of my work
- [x] I updated the PR name to follow our [PR naming guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/Pull-Request-Naming-Guide)
- [x] I ensured unit tests pass and wrote tests for new code
- [x] If working on UI, I checked and implemented accessibility (Dynamic Text and VoiceOver)
- [x] If adding telemetry, I read the [data stewardship requirements](https://github.com/mozilla-mobile/firefox-ios/wiki/Adding-Glean-Telemetry-Events) and will request a data review
- [x] If needed, I updated documentation and added comments to complex code
- [x] If needed, I added a backport comment (example `@Mergifyio backport release/v120`)

### CODE REVIEW

File: BrowserKit/Sources/Common/Theming/ThemeColourPalette.swift [hunk 1]
```diff
@@ -52,6 +52,7 @@ public protocol ThemeColourPalette {
     var actionPrimaryHover: UIColor { get }
     var actionPrimaryDisabled: UIColor { get }
     var actionSecondary: UIColor { get }
+    var actionSecondaryDisabled: UIColor { get }
```
Comment (cyndichin) [inline BrowserKit/Sources/Common/Theming/ThemeColourPalette.swift:55]: @FilippoZazzeroni does this already exist in our design system or should we follow up with the Acorn team?

File: BrowserKit/Sources/SummarizeKit/SummarizeViewModel.swift [hunk 2]
```diff
@@ -7,14 +7,17 @@ import ComponentLibrary
...
     let loadingA11yLabel: String
     let loadingA11yId: String
     let tabSnapshotA11yLabel: String
     let tabSnapshotA11yId: String
-    let summarizeTextViewA11yLabel: String
     let summarizeTextViewA11yId: String
     let brandLabel: String
+    let brandLabelA11yId: String
+    let brandImage: UIImage?
+    let brandImageA11yId: String
```
Comment (cyndichin) [inline BrowserKit/Sources/SummarizeKit/SummarizeViewModel.swift]: to make this cleaner, can we make another struct, maybe `BrandConfiguration` so that its clear that these are all used for one specific UI view?

File: BrowserKit/Sources/SummarizeKit/SummarizeViewModel.swift [hunk 3]
```diff
@@ -7,14 +7,17 @@ import ComponentLibrary
 
 public struct SummarizeViewModel {
     let titleLabelA11yId: String
+    let compactTitleLabelA11yId: String
     let loadingLabel: String
     let loadingA11yLabel: String
     let loadingA11yId: String
     let tabSnapshotA11yLabel: String
     let tabSnapshotA11yId: String
-    let summarizeTextViewA11yLabel: String
```
Comment (cyndichin) [inline BrowserKit/Sources/SummarizeKit/SummarizeViewModel.swift:15]: nice clean up!

File: BrowserKit/Sources/SummarizeKit/SummaryView/SummaryBrandCell.swift [hunk 4]
```diff
@@ -0,0 +1,81 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/
+
+import Common
+import Foundation
+import UIKit
+
+class SummaryBrandCell: UITableViewCell, ReusableCell, ThemeApplicable {
```
Comment (cyndichin) [inline BrowserKit/Sources/SummarizeKit/SummaryView/SummaryBrandCell.swift]: It seems we're adding a brand cell, in addition to the title animation. It would be helpful in the future for reviewers if we kept PRs small and separated these work from each other since they seem independent.

```suggestion
final class SummaryBrandCell: UITableViewCell, ReusableCell, ThemeApplicable {
```

File: BrowserKit/Sources/SummarizeKit/SummaryView/SummaryBrandCell.swift [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    private struct UX {
+        static let logoSize: CGFloat = 16.0
+        static let hPadding: CGFloat = 6.0
```
Comment (cyndichin) [inline BrowserKit/Sources/SummarizeKit/SummaryView/SummaryBrandCell.swift]: nit: 

```suggestion
        static let horizontalPadding: CGFloat = 6.0
```

File: firefox-ios/Client/Coordinators/SummarizeCoordinator.swift [hunk 6]
```diff
@@ -89,16 +89,24 @@ class SummarizeCoordinator: BaseCoordinator, SummarizerServiceLifecycle {
         } else {
             String(format: .Summarizer.HostedBrandLabel, AppName.shortName.rawValue)
         }
+        let brandImage: UIImage? = if summarizerNimbusUtils.isAppleSummarizerEnabled() {
+            UIImage(named: "appleIntelligence")
+        } else {
+            UIImage(named: "faviconFox")
+        }
```
Comment (cyndichin) [inline firefox-ios/Client/Coordinators/SummarizeCoordinator.swift:96]: this logic seems to make more sense in the view model, than the coordinator.

File: BrowserKit/Sources/SummarizeKit/SummaryView/SummaryView.swift [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+
+    // MARK: - ThemeApplicable
+    func applyTheme(theme: any Theme) {
+        self.theme = theme
+        compactTitleLabel.textColor = theme.colors.textPrimary
+        if #unavailable(iOS 26) {
+            closeButton.configuration?.baseBackgroundColor = theme.colors.actionTabActive
+        }
+        closeButton.configuration?.baseForegroundColor = theme.colors.textPrimary
+        tableView.reloadData()
```
Comment (cyndichin) [inline BrowserKit/Sources/SummarizeKit/SummaryView/SummaryView.swift]: it feels weird that we need to `reloadData` in the `applyTheme` method and may be inconsistent with how we use it elsewhere, why do we need this change?

File: BrowserKit/Sources/SummarizeKit/SummaryView/SummaryView.swift [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+        closeButton.accessibilityIdentifier = a11yId
+        closeButton.accessibilityLabel = a11yLabel
+    }
+
+    func showContent() {
+        tableView.alpha = 1.0
+    }
+
+    // MARK: - UIScrollViewDelegate
+    func scrollViewDidScroll(_ scrollView: UIScrollView) {
```
Comment (cyndichin) [inline BrowserKit/Sources/SummarizeKit/SummaryView/SummaryView.swift:196]: its a bit hard to follow this at first glance, could we add comments to clarify?

File: BrowserKit/Sources/SummarizeKit/SummarizeController.swift [hunk 9]
```diff
@@ -111,34 +121,7 @@ public class SummarizeController: UIViewController, Themeable, Notifiable, CAAni
...
-        $0.contentInset = UIEdgeInsets(
-            top: 0.0,
-            left: 0.0,
-            bottom: UX.tabSnapshotFinalPositionBottomPadding,
-            right: 0.0
-        )
-        $0.adjustsFontForContentSizeCategory = true
-        $0.isEditable = false
-    }
+    private let summaryView: SummaryView = .build()
```
Comment (cyndichin) [inline BrowserKit/Sources/SummarizeKit/SummarizeController.swift:124]: nice thanks for the refactor!

File: BrowserKit/Sources/SummarizeKit/SummarizeController.swift [hunk 10]
```diff
@@ -194,66 +172,26 @@ public class SummarizeController: UIViewController, Themeable, Notifiable, CAAni
...
         tabSnapshotContainer.accessibilityIdentifier = viewModel.tabSnapshotA11yId
         tabSnapshotContainer.accessibilityLabel = viewModel.tabSnapshotA11yLabel
 
-        closeButton.accessibilityIdentifier = viewModel.closeButtonModel.a11yIdentifier
-        closeButton.accessibilityLabel = viewModel.closeButtonModel.a11yLabel
-        closeButton.setImage(UIImage(named: StandardImageIdentifiers.Large.cross)?.withRenderingMode(.alwaysTemplate),
-                             for: .normal)
-        closeButton.addAction(
-            UIAction(handler: { [weak self] _ in
+        summaryView.configureCloseButton(
```
Comment (cyndichin) [inline BrowserKit/Sources/SummarizeKit/SummarizeController.swift]: torn between this, it's cleaner if we add these directly to when we initialize the close button, but then we won't be injecting it. Not sure what value we provide by injecting these over adding it in the initializer because this isn't a reusable component. 🤔 What do you think?

File: BrowserKit/Sources/SummarizeKit/SummarizeController.swift [hunk 11]
```diff
@@ -344,6 +269,39 @@ public class SummarizeController: UIViewController, Themeable, Notifiable, CAAni
...
+
+    private func summarize() {
+        loadingLabel.alpha = 1.0
+        errorView.alpha = 0.0
+        Task { [weak self] in
+            await self?.summarizeTask()
+        }
+    }
+
+    private func summarizeTask() async {
```
Comment (cyndichin) [inline BrowserKit/Sources/SummarizeKit/SummarizeController.swift:284]: I still don't like how we have this in the view controller, but not related to this ticket. 😅

File: BrowserKit/Sources/SummarizeKit/SummarizeController.swift [hunk 12]
```diff
@@ -353,23 +311,38 @@ public class SummarizeController: UIViewController, Themeable, Notifiable, CAAni
...
-        ###### \(viewModel.brandLabel)
-
         \(summary)
 
         ##### \(viewModel.summaryNote)
         """
-        titleLabel.text = webView.title
-        titleLabel.largeContentTitle = webView.title
-        summaryView.attributedText = parse(markdown: brandedSummary)
+        let inset = UIEdgeInsets(
```
Comment (cyndichin) [inline BrowserKit/Sources/SummarizeKit/SummarizeController.swift]: can we break this into its own method with a good description to make this method less verbose

File: BrowserKit/Sources/SummarizeKit/SummarizeController.swift [hunk 13]
```diff
@@ -353,23 +311,38 @@ public class SummarizeController: UIViewController, Themeable, Notifiable, CAAni
...
+                title: webView.title,
+                titleA11yId: viewModel.titleLabelA11yId,
+                compactTitleA11yId: viewModel.compactTitleLabelA11yId,
+                brandIcon: viewModel.brandImage,
+                brandIconA11yId: viewModel.brandImageA11yId,
+                brandName: viewModel.brandLabel,
+                brandNameA11yId: viewModel.brandLabelA11yId,
+                summary: parse(markdown: brandedSummary),
+                summaryA11yId: viewModel.summarizeTextViewA11yId,
+                contentOffset: inset
```
Comment (cyndichin) [inline BrowserKit/Sources/SummarizeKit/SummarizeController.swift]: what do you think about consolidating these into configurations so its easier to understand what each field belongs too. Also, we don't need to pass in so many different fields?

File: BrowserKit/Sources/SummarizeKit/SummaryView/SummaryTitleCell.swift [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+        fatalError("init(coder:) has not been implemented")
+    }
+
+    private func setup() {
+        contentView.addSubview(titleLabel)
+        NSLayoutConstraint.activate([
+            titleLabel.topAnchor.constraint(equalTo: contentView.topAnchor),
+            titleLabel.leadingAnchor.constraint(equalTo: contentView.leadingAnchor),
+            titleLabel.trailingAnchor.constraint(equalTo: contentView.trailingAnchor),
+            titleLabel.bottomAnchor.constraint(equalTo: contentView.bottomAnchor, constant: -20)
```
Comment (cyndichin) [inline BrowserKit/Sources/SummarizeKit/SummaryView/SummaryTitleCell.swift]: add magic numbers to UX struct

**Review by cyndichin — COMMENTED** [review]
thanks for working on improving the UI and the refactor!! 🔥  I left some initial comments to consider, let me know what you think!

Also, for future PRs, it'll make it easier / faster to review if we kept PRs smaller. Maybe we could've broken it up into these 3 parts.
- Refactoring 
- Adding Brand Label
- Adding Title Animation

File: BrowserKit/Sources/Common/Theming/ThemeColourPalette.swift (hunk 1 above)
Comment (FilippoZazzeroni) [reply BrowserKit/Sources/Common/Theming/ThemeColourPalette.swift:55]: yes it is already in that design system

File: BrowserKit/Sources/SummarizeKit/SummarizeViewModel.swift (hunk 2 above)
Comment (FilippoZazzeroni) [reply BrowserKit/Sources/SummarizeKit/SummarizeViewModel.swift]: yes sure

File: BrowserKit/Sources/SummarizeKit/SummaryView/SummaryBrandCell.swift (hunk 4 above)
Comment (FilippoZazzeroni) [reply BrowserKit/Sources/SummarizeKit/SummaryView/SummaryBrandCell.swift]: Ok yes sounds good, I'm going to make a better separation for next PR

File: BrowserKit/Sources/SummarizeKit/SummaryView/SummaryView.swift (hunk 8 above)
Comment (FilippoZazzeroni) [reply BrowserKit/Sources/SummarizeKit/SummaryView/SummaryView.swift:196]: yes sure

File: BrowserKit/Sources/SummarizeKit/SummaryView/SummaryView.swift (hunk 7 above)
Comment (FilippoZazzeroni) [reply BrowserKit/Sources/SummarizeKit/SummaryView/SummaryView.swift]: I saw it we are using it also elsewhere, honestly it makes sense to me as we cannot inject the theme into the table view directly so after a theme as being applied we need to refresh the table view so the cell have the updated theme. I don't know what could be the change, do u have any suggestions ?

File: BrowserKit/Sources/SummarizeKit/SummarizeController.swift (hunk 10 above)
Comment (FilippoZazzeroni) [reply BrowserKit/Sources/SummarizeKit/SummarizeController.swift]: yes super agree, I'm going to move evrything in the initializer

File: BrowserKit/Sources/SummarizeKit/SummarizeController.swift (hunk 12 above)
Comment (FilippoZazzeroni) [reply BrowserKit/Sources/SummarizeKit/SummarizeController.swift]: sure 😄

File: BrowserKit/Sources/SummarizeKit/SummarizeController.swift (hunk 13 above)
Comment (FilippoZazzeroni) [reply BrowserKit/Sources/SummarizeKit/SummarizeController.swift]: yes sure

File: BrowserKit/Sources/SummarizeKit/SummarizeController.swift (hunk 11 above)
Comment (FilippoZazzeroni) [reply BrowserKit/Sources/SummarizeKit/SummarizeController.swift:284]: ok lets discuss this, I'm open to any suggestions. I'm going to make a ticket to refactor it

File: firefox-ios/Client/Coordinators/SummarizeCoordinator.swift (hunk 6 above)
Comment (FilippoZazzeroni) [reply firefox-ios/Client/Coordinators/SummarizeCoordinator.swift:96]: yes I agree. I had a deeper look and also brand name should be move in the view model, but that would require that the model knows about `isAppleSummarizerEnabled`.  I'd prefer to keep those configuration in the coordinator so we can keep the view model free from knowing about nimbus configuration, what do you think ?

File: BrowserKit/Sources/SummarizeKit/SummarizeController.swift (hunk 10 above)
Comment (FilippoZazzeroni) [reply BrowserKit/Sources/SummarizeKit/SummarizeController.swift]: @cyndichin I tried to move it to the initializer of `SummaryView` and unfortunately it doesn't work as we can't pass the callback in the initializer given it is constrained to the controller's instance and it'd result in a compiler error that we are accessing self before full initialization.

File: BrowserKit/Sources/Common/Theming/ThemeColourPalette.swift (hunk 1 above)
Comment (cyndichin) [reply BrowserKit/Sources/Common/Theming/ThemeColourPalette.swift:55]: great! thanks for confirming!

File: BrowserKit/Sources/SummarizeKit/SummaryView/SummaryBrandCell.swift (hunk 4 above)
Comment (cyndichin) [reply BrowserKit/Sources/SummarizeKit/SummaryView/SummaryBrandCell.swift]: thank you!

File: BrowserKit/Sources/SummarizeKit/SummarizeController.swift (hunk 11 above)
Comment (cyndichin) [reply BrowserKit/Sources/SummarizeKit/SummarizeController.swift:284]: It depends if we decide to go with Redux route or stay with view model (maybe to be discussed with the team). At least for what we have now, maybe we should at least throw it in the view model and can be more testable? Would also be nice to have a test for this view controller as well to track memory leaks at the minimum.

File: firefox-ios/Client/Coordinators/SummarizeCoordinator.swift (hunk 6 above)
Comment (cyndichin) [reply firefox-ios/Client/Coordinators/SummarizeCoordinator.swift:96]: Thinking off the top of my head, In Redux, I feel that those configurations should live in the middleware and then we can update the state of the summarizer with the proper values. In this case, I feel that the coordinator should only need to know presentation logic and not the business logic. To me it bloats the coordinator a bit with this logic that feels like it belongs somewhere else. It seems we need to know about the nimbus checks for making the summarize service (model layer) and the brand label (view model layer) and not specific to presentation.

If we decide to keep the view model architecture, then this is how I interpret the flow where we have the Coordinator -> View Controller -> View Model -> Model. So the nimbus configurations looks very closely related to View Model / Model area in which we setup the service and as well determine the brand label over the coordinator. 

I did see this in our wiki (below), but I'm wondering what the A/B test variants refer to, but the summarizer variants should be easy to access anywhere. 
> Coordinator should control the flow in our app, which means creating VCs and view models and calling the router to present and push. It shouldn’t decide what to show or contain business logic. The one exception for this is that coordinator can contain A/B tests variants since it’s convenient.

Maybe can get @lmarceau thoughts as well? 

image

Sorry for the long comment, definitely don't need this to block this PR and can do a follow refactor, but this is my first thought looking at this code. Let me know what you think!

File: BrowserKit/Sources/SummarizeKit/SummaryView/SummaryView.swift (hunk 7 above)
Comment (cyndichin) [reply BrowserKit/Sources/SummarizeKit/SummaryView/SummaryView.swift]: Oh okay, I thought since we have `listenForThemeChanges` which is supposedly supposed to be called after a theme change and be reflected in our subviews, that it would be handled. 

Also, I notice we don't apply the theme in `configureSummaryView`, should we add `summaryView.applyTheme` in the configuration as well?

If we still need to reload then adding a comment would be great!

File: BrowserKit/Sources/SummarizeKit/SummarizeController.swift (hunk 10 above)
Comment (cyndichin) [reply BrowserKit/Sources/SummarizeKit/SummarizeController.swift]: okay makes sense, what about adding it to the configurations in summary view instead so we contain the close button logic in one area?

**Comment by FilippoZazzeroni** [comment]
Thanks a lot Cyndi.  Happy to chat live too. I think Redux as we spoke already is nice to handle the 3 entry points and plus we can use it to provide the state for the Library and thus moving the configuration from the Coordinator to a SummaryState structure.
Yes `listenToThemeChanges` should reload all the subviews when there is a theme change but not when we setup the UI the first time. Tho I can I move it elsewhere and remove it from the applyTheme.
The `applyTheme` for the summary view is already in the apply theme in the `SummarizeController`. 
Regarding the `configureCloseButton` unfortunately we can't have one method so far, cause the close button is shown before the summary is present and thus needs to be already configured. I can refactor it tho to move the configuration in one place and just have method to pass the computed summary.

**Comment by FilippoZazzeroni** [comment]
Ok @cyndichin I managed to remove `reloadData()` 🎉  since the applyTheme on the controller runs at view did load, when the summary is configured the view has already a theme applied so the collection view builds with the correct style.

**Comment by cyndichin** [comment]
> Ok @cyndichin I managed to remove `reloadData()` 🎉 since the applyTheme on the controller runs at view did load, when the summary is configured the view has already a theme applied so the collection view builds with the correct style.

Nice!!! thank you

**Comment by FilippoZazzeroni** [comment]
@mergify backport release/v143.0
